### PR TITLE
Evolution of push rules

### PR DESCRIPTION
**Pull request by dbkr**
Add enable/disable overlay for push rules, with API. Give server default rules rule_ids.

Contains a couple of lines that are over the threshold for the PEP8 warning: not sure what our style is on lines like these.

### CODE REVIEW

**Comment by ara4n** [comment]
common sense trumps line length on pep8 imo.

**Comment by NegativeMjark** [comment]
Does get_push_rules_enabled_for_user_name take a user_id string?
If so it might be better to call it get_push_rules_enabled_for_user for consistency with the other database apis?

**Comment by dbkr** [comment]
NegativeMjark: Okay - I'd put user_name to mark them as taking a user id string rather than a user object, but I see that doesn't match precedent.

File: synapse/storage/push_rule.py [hunk 1]
```diff
@@ -46,6 +46,28 @@ def get_push_rules_for_user_name(self, user_name):
...
+        )
+
+    @defer.inlineCallbacks
+    def get_push_rule_enabled_by_user_rule_id(self, user_name, rule_id):
+        results = yield self._simple_select_list(
+            PushRuleEnableTable.table_name,
+            {'user_name': user_name, 'rule_id': rule_id},
+            ['enabled']
+        )
+        if len(results) == 0:
```
Comment (erikjohnston) [inline synapse/storage/push_rule.py]: We usually just use `if not results:` for things like this.

File: synapse/storage/push_rule.py [hunk 2]
```diff
@@ -216,3 +252,13 @@ class PushRuleTable(Table):
...
+
+
+class PushRuleEnableTable(Table):
+    table_name = "push_rules_enable"
+
+    fields = [
+        "user_name",
+        "rule_id",
+        "enabled"
+    ]
```
Comment (erikjohnston) [inline synapse/storage/push_rule.py]: We don't generally use `Table` anymore, instead we just hard code table name directly in the SQL.

**Comment by erikjohnston** [comment]
Otherwise LGTM.

**Comment by erikjohnston** [comment]
LGTM.

**Comment by erikjohnston** [comment]
Actually, can you move the delta script into delta/14?

**Pull request merged**